### DISCUSSION TITLE
Feature Request: 'something' that would allow the order of lines read from files (in particular, CSV) to be knowable/reconstructible

### DISCUSSION THREAD

**Discussion by rpbouman** [duckdb/duckdb#12882] [Ideas]
I noticed the CSV reader tends to return rows in order that corresponds to the position of the lines in the file.
Initially that seemed natural to me, but given the fact that the CSV reader could have parallelism enabled, it suddenly hit me that this may not be so natural at all.

I'm using the CSV reader for processing raw text files (books), and it's pretty essential that I can keep the lines in the same order of appearance as in the file.

Now, I realize that the primary use case for the reader is turning data files into relations, and that as such I should never expect to rely on the physical position of the rows. But I figured it might be worth asking for a feature that would allow the order to be preserved, or for a bit of data to be exposed that would allow the order to be reconstructed.

For example:
- I can imagine an flag argument for read_csv that would simply be `preserve_line_order = TRUE|FALSE`. Advantage would be that it would be very easy to use, and also give maximum power for the internals to decide how to do it.
- I can also imagine the CSV reader exposing the row_index (or if its easier to implement, the row_offset as byte position) for each row as a column (kinda like how the file_name column exposes the file from whence the row came). This would give the receiver the possibility to explicitly order on file_name, row_index. Perhaps in case of parallel CSV reading, another field that conveys the offset used by that thread would also be necessary to fully reconstruct the original line order.
- Another approach would be to have a a separate line reader tailored for this purpose. Disadvantage would be that it would be separate, new component, advantage could be that it maybe could befaster than the CSV reader, since it would only need to read lines, and not worry about columns, quoting, escaping, data types etc.

**Comment by rpbouman** [comment]
@hawkfish just explained to me on discord that adding `ROW_NUMBER() OVER ()` would ensure the order:

> You can generate the ordering column on ingestion by using a window function:
> SELECT *, row_number() OVER() AS rowid
> FROM read_csv(...)
> 
> This will trigger our streaming window operator instead of the materialising one, so the ingestion will still be single-threaded, but the resulting data will have the row numbers you want. 
 
I'm grateful for that suggestion, but it doesn't seem self-evident that this behavior should always be guaranteed either. 
If it is, it is not documented.

I suppose it's a matter of the level at which point ROW_NUMBER() considers the data to be a relational set. It seems reasonable to me for ROW_NUMBER to do that once the rows are produced by the CSV reader, and if that is so, it would take the result of the reader as a given, and not care about whether parallel csv reading was used or not. 

Whether there is a row-number to line-number correspondence seems something that should be the business of the csv reader, and if that is so, and if it should be controllable, then it should be via the options of the csv reader. 

Anyway  - for now I'm thankful for @hawkfish' suggestion, I can move on for now.

**Comment by rpbouman** [comment]
@pdet just confirmed to me that the CSV reader already preserves line order - even when parallelism is enabled:

> Hey Roland, no worries, happy to answer the question :-).
>  The CSV Scanner should guarantee the row order, unless the preserve_insertion_orderis set to false.
> Roland Bouman — Today at 5:13 PM
> That's awesome! Even in case of parallism enabled?
> pdet — Today at 5:13 PM
> Yes, even if parallelism is enabled
> 

So - that's just awesome! Very happy!

I will add a issue to duckdb-web to get this behavior documented. 
I'll pick up that issue, and I'll close this discussion once it lands in the manual.

Thanks all!

**Comment by rpbouman** [comment]
The order-preserving behavior of the CSV reader is now documented. From my perspective this discussion can be closed as there is no need for any new feature.

For reference:
https://github.com/duckdb/duckdb-web/pull/3273/commits

https://github.com/duckdb/duckdb-web/issues/3266
https://github.com/duckdb/duckdb-web/issues/3249

**Comment by soerenwolfers** [comment]
@pdet @hawkfish  Do you happen to know how to parquet reader behaves as well? 
@szarnyasg I think the docs at https://duckdb.org/docs/sql/dialect/order_preservation could be a bit more explicit about what it means to "preserve insertion order". Insertion of what into what? What if files are being read in a CTE and `row_number() over ()`ed elsewhere?

**Comment by soerenwolfers** [comment]
Personally, I believe giving the various scanners `line_number` (or `row_index`) arguments would be the cleanest and most convenient solution.

**Comment by rpbouman** [comment]
Hi Soeren, 

I do not have an answer to these questions, 

However there's currently at least one issue with the CSV reader: https://github.com/duckdb/duckdb/issues/12933

If you're worried about a specific use case, my suggestion would be to try and come up with a reproducible example that violates the principle, and then ask informally on discord whether that in fact constitutes a bug, and then go from there.

Whether or not a line_number is the cleanest solution will, I guess, depend on all kinds of implementation details, as well as specific use cases (for example, what about hive partitioned directories). 

I think the best way forward is to ask what components of duckdb are expected/supposed to adhere to the principle of preserving order, document those cases, and then, file bugs whenever you find violations.
